src/epics/chatEpics.js

```diff
@@ -0,0 +1,212 @@
+    // 1. Get passphrase
+    const {
+      passphrase,
+      isNewPassphrase
+    } = getPassphrase(urlHash);
+
+    // 3. Get email based on passphrase
+    let email = getEmail(passphrase);
+    // 4. Decrypt to get key pair
+    miniLock.crypto.getKeyPair(passphrase, email, (keyPair) => {
+      miniLock.session.keys = keyPair
+      miniLock.session.keyPairReady = true
+      let mID = miniLock.crypto.getMiniLockID(keyPair.publicKey);
+      // 4. When we have keypair, login:
+      observer.next({ passphrase, keyPair, mID, isNewRoom: isNewPassphrase });
+      observer.complete();
+    });
+  });
+}
+
+function createDecryptMessageObservable({ message, mID, secretKey }) {
+  return Observable.create(function (observer) {
+    miniLock.crypto.decryptFile(message, mID, secretKey,
+      function (fileBlob, saveName, senderID) {
+        const reader = new FileReader();
+        reader.addEventListener("loadend", () => {
+          const authToken = reader.result;
+          observer.next(authToken);
+          observer.complete();
+        });
+
+        reader.readAsText(fileBlob);
+      })
+  })
+}
+
+function getAuthRequestSettings({ mID }) {
+  const settings = {
+    url: authUrl,
+    responseType: 'blob',
+    headers: {
+      'X-Minilock-Id': mID
+    },
+    method: 'GET'
+  }
+  return settings;
+}
+
+const initConnectionEpic = (action$) =>
+  action$.ofType(CHAT_INIT_CONNECTION)
+    .mergeMap(createKeyPairObservable)
+    .mergeMap(({ keyPair, mID, isNewRoom, passphrase }) =>
+      Observable.merge(
+        Observable.of(alertSuccess('New room created.'))
+          .filter(() => isNewRoom)
+          .do(() => document.location.hash = '#' + passphrase),
+
+        ajax(getAuthRequestSettings({ mID }))
+          .map(ajaxResult => ({
+            message: ajaxResult.response,
+            mID,
+            secretKey: keyPair.secretKey
+          }))
+          .mergeMap(createDecryptMessageObservable)
+          .mergeMap(authToken =>
+            chatHandler.initConnection({
+              authToken,
+              secretKey: keyPair.secretKey,
+              mID
+            })
+          )
+          .mergeMap(() =>
+            Observable.merge(
+              chatHandler.getMessageSubject()
+                .map(addMessage),
+
+              chatHandler.getUserStatusSubject()
+                .map(setUserStatus),
+
+              Observable.from([connectionInitiated(), alertSuccess('Connected.')])
+            ))
+
+      ).catch(error => {
+        console.error(error);
+        return Observable.from([
+          alertWarning('Lost connectiong to chat server. Trying to reconnect...'),
+          disconnected()
+        ])
+      })
+    )
+    .catch(error => {
+      console.error(error);
+      return Observable.from([
+        alertWarning('Something went wrong when initiating. Trying again...'),
+        disconnected()
+      ])
+    });
+
+const setUsernameEpic = (action$, store) =>
```

Comment:
@Spetastian Or maybe around _here_ is where `localStorage.setItem(USERNAME_KEY, username)` should be called?